tests/lib/rules/no-unreachable-loop.js

```diff
@@ -0,0 +1,430 @@
+        "async function foo() { for await (const a of b) <body> }"
+    ]
+};
+
+const validLoopBodies = [
+    ";",
+    "{}",
+    "{ bar(); }",
+    "continue;",
+    "{ continue; }",
+    "{ if (foo) break; }",
+    "{ if (foo) { return; } bar(); }",
+    "{ if (foo) { bar(); } else { break; } }",
+    "{ if (foo) { continue; } return; }",
+    "{ switch (foo) { case 1: return; } }",
+    "{ switch (foo) { case 1: break; default: return; } }",
+    "{ switch (foo) { case 1: continue; default: return; } throw err; }",
+    "{ try { return bar(); } catch (e) {} }",
+
+    // unreachable break
+    "{ continue; break; }",
+
+    // functions in loops
+    "() => a;",
+    "{ () => a }",
+    "(() => a)();",
+    "{ (() => a)() }",
+
+    // loops in loops
+    "while (a);",
+    "do ; while (a)",
+    "for (a; b; c);",
+    "for (; b;);",
+    "for (; ; c) if (foo) break;",
+    "for (;;) if (foo) break;",
+    "while (true) if (foo) break;",
+    "while (foo) if (bar) return;",
+    "for (a in b);",
+    "for (a of b);"
+];
+
+const invalidLoopBodies = [
+    "break;",
+    "{ break; }",
+    "return;",
+    "{ return; }",
+    "throw err;",
+    "{ throw err; }",
+    "{ foo(); break; }",
+    "{ break; foo(); }",
+    "if (foo) break; else return;",
+    "{ if (foo) { return; } else { break; } bar(); }",
+    "{ if (foo) { return; } throw err; }",
+    "{ switch (foo) { default: throw err; } }",
+    "{ switch (foo) { case 1: throw err; default: return; } }",
+    "{ try { return bar(); } catch (e) { break; } }",
+
+    // unreachable continue
+    "{ break; continue; }",
+
+    // functions in loops
+    "{ () => a; break; }",
+    "{ (() => a)(); break; }",
+
+    // loops in loops
+    "{ while (a); break; }",
+    "{ do ; while (a) break; }",
+    "{ for (a; b; c); break; }",
+    "{ for (; b;); break; }",
+    "{ for (; ; c) if (foo) break; break; }",
+    "{ for(;;) if (foo) break; break; }",
+    "{ for (a in b); break; }",
+    "{ for (a of b); break; }",
+
+    /**
+     * Additional cases where code path analysis detects unreachable code: after loops that don't have a test condition or have a
+     * constant truthy test condition, and at the same time don't have any exit statements in the body. These are special cases
+     * where this rule reports error not because the outer loop's body exits in all paths, but because it has an infinite loop
+     * inside, thus it (the outer loop) cannot have more than one iteration.
+     */
+    "for (;;);",
+    "{ for (var i = 0; ; i< 10) { foo(); } }",
+    "while (true);"
+];
+
+/**
+ * Creates source code from the given loop template and loop body.
+ * @param {string} template Loop template.
+ * @param {string} body Loop body.
+ * @returns {string} Source code.
+ */
+function getSourceCode(template, body) {
+    const loop = template.replace("<body>", body);
+
+    return body.includes("return") && !template.includes("function") ? `function someFunc() { ${loop} }` : loop;
```

Comment:
just want to point out: `string.include()` is not a reliable way to do the check, e.g.
```js
let s = "return 1;";
// or
arr.map(function (){return 1;});
```
But it seems no harm wrapping these cases in function bodies. 😄

Reply:
Agreed, this is far from ideal, but couldn't figure out a better way to test all combinations (and not add too much complexity at the same time).